File: app/src/main/java/com/onwordiesquire/mobile/marvelapp/characterlookup/presentation/SearchPresenter.java
package com.onwordiesquire.mobile.marvelapp.characterlookup.presentation;

import com.onwordiesquire.mobile.marvelapp.characterlookup.domain.usecase.GetCharacter;
import com.onwordiesquire.mobile.marvelapp.data.CharacterDataRepository;

import javax.inject.Inject;

import rx.subscriptions.CompositeSubscription;
import timber.log.Timber;

/**
 * Created by michelonwordi on 10/24/16.
 */
public class SearchPresenter implements CharacterLookUpContract.Presenter{

    GetCharacter getCharacter;
    CompositeSubscription compositeSubscription;
    private CharacterLookUpContract.View mvpView;

    @Inject
    public SearchPresenter(GetCharacter getCharacter) {
        this.getCharacter = getCharacter;
    }

    @Override
    public void attachView(CharacterLookUpContract.View view) {
        this.mvpView = view;
        compositeSubscription = new CompositeSubscription();
    }

    @Override
    public void detachView() {
        this.mvpView = null;
        compositeSubscription.clear();
    }

    public boolean isViewAttached()
    {
        return mvpView != null;
    }

    @Override
    public void loadCharacter(String name) {
        if(isViewAttached())
        {
            compositeSubscription.add(
                    getCharacter.buildObservable(new GetCharacter.RequestValues(null,name))
                    .subscribe(responseValues -> {
                        mvpView.showProgressIndicator(false);
                        mvpView.showCharacterDetails(responseValues.getCharacter());
                    },e->{
                        if(e instanceof CharacterDataRepository.EmptyResultsException)
                        {
                            mvpView.showProgressIndicator(false);
                            mvpView.showEmptyState("An incorrect name was specified");

                        }else{
                            mvpView.showProgressIndicator(false);
                            mvpView.showError("An error has occurred");
                            Timber.e(e.getMessage(),e);
                        }
                    })
            );
        }


    }

    @Override
    public void loadPreviousSearchTerms() {

    }


//    public void loadLastSearchedCharacters() {
//
//        compositeSubscription.add(dataManager.getRecentSearches()
//                .subscribeOn(Schedulers.io())
//                .observeOn(AndroidSchedulers.mainThread())
//                .flatMap(Observable::from)
//                .distinct()
//                .take(5)
//                .toList()
//                .subscribe(recentSearches -> {
//                    getMvpView().displayLastFiveSearches(recentSearches);
//                },e->{
//                    getMvpView().showError("An Error Happened");
//                    Timber.e(e.getMessage(),e);
//                })
//        );
//
//    }

//    public void loadCharacter(String name, String apikey, String timestamp, String privateKey) {
//        isViewAttached();
//        getMvpView().showProgressIndicator(true);
//
//        compositeSubscription.add(dataManager.getCharacter(name, apikey, timestamp, privateKey)
//                .subscribeOn(Schedulers.io())
//                .observeOn(AndroidSchedulers.mainThread())
//                .subscribe(characterData -> {
//                    getMvpView().showProgressIndicator(false);
//                    getMvpView().showCharacterDetails(characterData);
//                }, e -> {
//                    if (e instanceof CharacterDataRepositoryImpl.EmptyResultsException) {
//                        getMvpView().showProgressIndicator(false);
//                        getMvpView().showEmptyState();
//                    } else {
//                        getMvpView().showProgressIndicator(false);
//                        getMvpView().showError("An Error Happened");
//                        Timber.e(e.getMessage(),e);
//
//                    }
//                }));
//    }
}

File: app/src/main/java/com/onwordiesquire/mobile/marvelapp/presentation/characterSearch/refactor/SearchPresenter.kt
package com.onwordiesquire.mobile.marvelapp.presentation.characterSearch.refactor

import com.onwordiesquire.mobile.marvelapp.data.refactor.MarvelDataManager
import com.onwordiesquire.mobile.marvelapp.presentation.BasePresenter
import com.onwordiesquire.mobile.marvelapp.presentation.MvpView
import com.onwordiesquire.mobile.marvelapp.util.EmptyResultsException
import com.onwordiesquire.mobile.marvelapp.util.VISIBLE
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import rx.subscriptions.CompositeSubscription
import timber.log.Timber
import javax.inject.Inject

/**
 * Created by michel.onwordi on 10/07/2017.
 */
class SearchPresenter @Inject constructor(val dataManager: MarvelDataManager) : BasePresenter<SearchView>() {

    override fun attachView(view: SearchView) {
        super.attachView(view)
        compositeSubscription = CompositeSubscription()
    }

    lateinit var compositeSubscription: CompositeSubscription

    override fun detachView() {
        super.detachView()
        compositeSubscription.clear()
    }

    fun findCharacter(name: String, timeStamp: String) {
        mvpView?.let {
            it.showProgress(VISIBLE)
            dataManager.getCharacter(name, timeStamp)
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe({ character ->
                        it.showProgress(!VISIBLE)
                        it.showCharacterDetails(character)
                    }, { error ->
                        it.showProgress(!VISIBLE)
                        if (error is EmptyResultsException) {
                            it.showEmptyState("Sorry can't find that name")
                        } else {
                            it.showError("An Error Happened")
                            Timber.e(error.message, error)
                        }
                    })
        }

    }

    fun onSearchClick(textValue: String) =
            if (textValue.isEmpty()) {
                Timber.i("Search field Not empty")

            } else {

                Timber.i("Search field  empty")
            }
}